### PR TITLE
add liveness probe to handler

### PR DESCRIPTION
**Pull request by elliott-beach**
Signed-off-by: Elliott Beach <elliott2.71828@gmail.com>

This adds a simple health check to the kubernetes definition. using the default parameters for timeouts etc.

Converting this to an http-check  on `:10254/healthz` would do what you want, as then the actual checking would be done in the watchdog process. Should that be done?

### CODE REVIEW

**Comment by e-beach** [comment]
This should also be changed to use `tempdir()` rather than `tmp`.

**Comment by alexellis** [comment]
Thanks @e-beach - this looks like parity with Docker Swarm. We should have an environmental variable that can disable the addition of this probe into all specs. I.e. `enable_function_readiness_probe` - true by default.

File: handlers/deploy.go [hunk 1]
```diff
@@ -136,6 +136,13 @@ func makeDeploymentSpec(request requests.CreateFunctionRequest) *v1beta1.Deploym
 								},
 							},
 							ImagePullPolicy: v1.PullAlways,
+							LivenessProbe: &apiv1.Probe{
+								Handler: apiv1.Handler{
+									Exec: &apiv1.ExecAction{
```
Comment (alexellis) [inline handlers/deploy.go]: I assume this creates some defaults for time until first run and time between repeating the check? For maintainability we should create the values explicitly

**Comment by alexellis** [comment]
Had a chance to test this out last night, seems to work as expected.

File: handlers/deploy.go (hunk 1 above)
Comment (e-beach) [reply handlers/deploy.go]: I set the value for the initial delay to 10, but couldn't figure out what the default value is.

**Comment by e-beach** [comment]
Added the environment variable, not tested.

**Comment by alexellis** [comment]
>not tested.

🙀

File: handlers/deploy.go [hunk 2]
```diff
@@ -85,8 +89,23 @@ func MakeDeployHandler(functionNamespace string, clientset *kubernetes.Clientset
...
-func makeDeploymentSpec(request requests.CreateFunctionRequest) *v1beta1.Deployment {
+func makeDeploymentSpec(request requests.CreateFunctionRequest, enableProbe bool) *v1beta1.Deployment {
 	envVars := buildEnvVars(request)
+	probe := &apiv1.Probe{
+		Handler: apiv1.Handler{
+			Exec: &apiv1.ExecAction{
+				Command: []string{"cat", "/tmp/.lock"},
+			},
+		},
+		InitialDelaySeconds: 10,
```
Comment (alexellis) [inline handlers/deploy.go]: I guess initial could be lower, like `2` or `3`?

**Comment by e-beach** [comment]
@alexellis I'm wondering why I don't see output from here in the logs: https://github.com/openfaas/faas-netes/blob/2c29d1a8784e997a969bc9c064e0c13676b5a07d/server.go#L50-L51 
Got a clue?
Logs:
```
[I] ⋊> ~/m/p/faas-cli on master ⨯ kubectl logs faas-netesd-1855941297-mtx08                                                                                                                                                           22:27:07
2017/10/04 03:24:19 object is being deleted: deployments.extensions "nodejs-echo" already exists
2017/10/04 03:24:19 object is being deleted: deployments.extensions "shrink-image" already exists
2017/10/04 03:24:19 Created deployment - ruby-echo
2017/10/04 03:24:20 Created service - ruby-echo
2017/10/04 03:24:20 {"service":"ruby-echo","image":"alexellis/ruby-echo","network":"func_functions","envProcess":"ruby index.rb","envVars":null,"constraints":[]}
2017/10/04 03:24:20 Created deployment - url-ping
2017/10/04 03:24:20 Created service - url-ping
2017/10/04 03:24:20 {"service":"url-ping","image":"ebeach0/faas-url-ping","network":"func_functions","envProcess":"python index.py","envVars":null,"constraints":[]}

````

**Comment by e-beach** [comment]
Also, I noticed that if I `docker exec`d into the container and removed the lock file, kubernetes did not restart the container. 

I should think of a better way to test that the environment variable works, rather than examining that sort of low-level detail.

**Comment by alexellis** [comment]
I actually did the same thing and Kubernetes did terminate and re-schedule the container. I wonder if we are running it the same way?

File: handlers/deploy.go [hunk 3]
```diff
@@ -53,7 +57,7 @@ func MakeDeployHandler(functionNamespace string, clientset *kubernetes.Clientset
 			return
 		}
 
-		deploymentSpec := makeDeploymentSpec(request)
+		deploymentSpec := makeDeploymentSpec(request, config.EnableFunctionReadinessProbe)
```
Comment (alexellis) [inline handlers/deploy.go]: Let's pass the whole config object to each endpoint.

**Comment by alexellis** [comment]
This is shaping up nicely. 🚀

File: handlers/deploy.go (hunk 3 above)
Comment (e-beach) [reply handlers/deploy.go]: @alexellis Sure, but it's not clear to me why that's preferred. Would you mind explaining?

**Comment by e-beach** [comment]
@alexellis Travis build is failing, and it's not my fault! The last commits broke it, as the `--squash` option is not supported at all by the Travis docker build system.

Incidentally, a similar problem has been vexing for me: minikube doesn't support docker edge either.
That is,
```
eval (minikube docker-env) 
docker build . -t e-beach/faas-netes
```
results in an error due to the `--from` flag on `COPY` not being supported in the Dockerfile.

**Comment by alexellis** [comment]
Two things - we should enable experimental mode on Travis (I broke the Makefile). 2nd thing you can either build locally with multi-stage and push to the gun each time or use the non-multi Dockerfile. See Makefile for instructions

**Comment by alexellis** [comment]
Hey how is your manual testing looking now? Can you trigger a re-schedule by deleting the .lock file?

**Comment by e-beach** [comment]
My manual testing cycle has been tedious with this repo, will work on it now but could take a little time.

I should also add a unit test for the environment variable.

**Comment by alexellis** [comment]
>My manual testing cycle has been tedious with this repo

What does it look like? Can I help?

**Comment by e-beach** [comment]
I'll share what I'm doing in a `script` session.

**Comment by alexellis** [comment]
Btw did you see any formatting errors re: gofmt? Do you use vscode to edit?

```
Step 8/16 : RUN gofmt -l -d $(find . -type f -name '*.go' -not -path "./vendor/*")   && CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o faas-netes .
 ---> Running in 18e923f9666d
./handlers/deploy.go
diff ./handlers/deploy.go gofmt/./handlers/deploy.go
--- /tmp/gofmt062482354	2017-10-07 18:47:59.404842981 +0000
+++ /tmp/gofmt569191273	2017-10-07 18:47:59.404842981 +0000
@@ -5,12 +5,12 @@
 
 import (
 	"encoding/json"
-	"path/filepath"
 	"fmt"
 	"io/ioutil"
 	"log"
 	"net/http"
 	"os"
+	"path/filepath"
 	"regexp"
 
 	"github.com/openfaas/faas/gateway/requests"
```

**Comment by alexellis** [comment]
Building locally I got this error:

```
# github.com/openfaas/faas-netes/handlers
handlers/deploy.go:62: cannot use config (type *DeployHandlerConfig) as type DeployHandlerConfig in argument to makeDeploymentSpec
```

**Comment by e-beach** [comment]
Sorry, I need to test, yes will fix.

**Comment by e-beach** [comment]
Probably shouldn't have pushed.

**Comment by e-beach** [comment]
Currently debugging `minikube start` hanging, will fix.

**Comment by alexellis** [comment]
You could try deleting / rebooting. Do you use the vbox driver or xhyve?

**Comment by e-beach** [comment]
vbox, I found some related stackoverflow threads so let me read through them.

**Comment by e-beach** [comment]
Looks like it worked, thanks!

**Comment by alexellis** [comment]
This guide has a way of using xhyve which I think is a bit quicker - https://blog.alexellis.io/minikube-behind-proxy/ - try it some time when you have a moment.

**Comment by e-beach** [comment]
I was able to test that adding the handler worked (problem was I was using the wrong image :P).
I think as long as we can see the liveness check in the output of `kubernetes describe pod url-ping`, we know it worked.

```
Name:		url-ping-1330243947-pwpp2
Namespace:	default
Node:		minikube/192.168.99.100
Start Time:	Sat, 07 Oct 2017 14:14:34 -0500
Labels:		faas_function=url-ping
		pod-template-hash=1330243947
Annotations:	kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"default","name":"url-ping-1330243947","uid":"c0d67d5b-ab93-11e7-8c94-0800277c9ab6...
Status:		Running
IP:		172.17.0.8
Created By:	ReplicaSet/url-ping-1330243947
Controlled By:	ReplicaSet/url-ping-1330243947
Containers:
  url-ping:
    Container ID:	docker://720dc84a9418b6139703ba79299f581e39a502e32f4cc5559562e4b855755a7f
    Image:		ebeach0/faas-url-ping
    Image ID:		docker-pullable://ebeach0/faas-url-ping@sha256:b7c3d466ae6a26e3dcc3cef3237c8d46c65499cda60f21ff3f8cd84ad8a224db
    Port:		8080/TCP
    State:		Running
      Started:		Sat, 07 Oct 2017 14:14:35 -0500
    Ready:		True
    Restart Count:	0
    Liveness:		exec [cat /tmp/.lock] delay=3s timeout=1s period=10s #success=1 #failure=3
    Environment:
      fprocess:	python index.py
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-csdsp (ro)
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	True 
  PodScheduled 	True 
Volumes:
  default-token-csdsp:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-csdsp
    Optional:	false
QoS Class:	BestEffort
Node-Selectors:	<none>
Tolerations:	<none>

```

**Comment by e-beach** [comment]
I might open a PR on derek because that behavior is incorrect, even if I should squash my commits. I signed all of them. Problem seems to occur after rebasing.

**Comment by alexellis** [comment]
He is meant to check all commits in the PR.

**Comment by rgee0** [comment]
`clean up set_travis_experimental_mode` appears to not be signed.

**Comment by e-beach** [comment]
> He is meant to check all commits in the PR.

Sorry, I could have been more polite and inquisitive my comment. 
@rgee0 Thanks.
Going out with some friends now

**Comment by alexellis** [comment]
Signing is OK, but we need sign-off. I just type in `git commit -s -m "...."`

Third answer here might also be good for "auto sign-off" - https://stackoverflow.com/questions/15015894/git-add-signed-off-by-line-using-format-signoff-not-working

**Comment by e-beach** [comment]
@alexellis 
Added tests, and manually verified that `healthcheck` exists in the created pod by default, but not with `enable_function_readiness_probe` set to false.

Signing problem was related to minor git mistake during rebase.

>Third answer here might also be good for "auto sign-off" - https://stackoverflow.com/questions/15015894/git-add-signed-off-by-line-using-format-signoff-not-working

Unless I build git off of master, not ready quite yet.

**Comment by e-beach** [comment]
>So for the watchdog let's start by returning everything is OK once the lock
file is in place - then go from there. You'll also need to update
faas-netes to programmatically add the healthcheck and readiness probe to
new functions.


Also, there probably should be some more communication about the readiness probe.
This actually only implements the healthcheck, not the readiness probe, we should do both before merging.

@alexellis I'm not sure how "then go from there" translates into code, either.

**Comment by alexellis** [comment]
What you have is valuable so I'm OK to merge as soon as possible. It will prevent functions from being accessed before they've started up.

**Comment by elliotbeach** [comment]
Cool, this should be good to merge.

File: test/read_config_test.go [hunk 4]
```diff
@@ -60,3 +64,29 @@ func TestRead_ReadAndWriteTimeoutConfig(t *testing.T) {
 		t.Fail()
 	}
 }
+
+func TestRead_EnableFunctionReadinessProbeConfig(t *testing.T) {
```
Comment (alexellis) [inline test/read_config_test.go]: One more test please for the "default" which should be on.

Comment (elliott-beach) [reply test/read_config_test.go]: @alexellis fixed.

Comment (alexellis) [reply test/read_config_test.go]: Perfect 👍 really pleased with this work, good job

**Comment by alexellis** [comment]
Merged! Just need to document the new flag `enable_function_readiness_probe` somewhere in the OpenFaaS Kubernetes documentation. Probably the deployment guide?

**Comment by elliott-beach** [comment]
Sure. The timeouts flags should probably also be documented. Even though
they are the same for the faas gateway, I will check to see if it’s clear
that they can also be used for faas-netes.

**Pull request merged**